Implemented SlidingWindowWalker

magicDGS opened the pull request:
New implementation of `SlidingWindowWalker` with some ideas from the discussion in #1528.

The thinks that are requested in #1198 still holds, but now it is more general: padding option is added and construction of windows are done by interval.

The code contain a lot of TODO because it relies on changes implemented in #1567, and because it is suppose to be a walker over `ReadWindow` instead of `SimpleInterval`+`ReadsContext` if reads are available.

I think that with these changes it could be general to be extended by `ReadWindowWalker` and by users that needs a different way of "slide" over intervals.

magicDGS commented:
I was thinking after checking your `ReadWindowWalker` implementation and the HaplotypeCaller branch that with this class we can do a general `SlidingWindowWalker`, which could be extended by `ReadWindowWalker` or it could be possible that `HaplotypeCaller` directly implements `SlidingWindowWalker`.

@droazen, could you have a look and tell me what do you think about the idea?

droazen commented:
@magicDGS The HaplotypeCaller traversal has undergone some changes in the past few weeks to improve performance and bring the output of the tool closer to GATK3. There is now an `AssemblyRegionWalker` that divides the intervals into active and inactive regions, in a greatly simplified version of the GATK3 traversal.

Initially, I did plan on having `AssemblyRegionWalker` extend the former `ReadWindowWalker`, or an adapted version of your `SlidingWindowWalker`, and I did implement it like this at first, but ultimately I collapsed it into a single class for several reasons:

* Inheriting from a more generic traversal type caused usability issues and confusion with respect to the command-line arguments. The `ReadWindow` was the unit of processing for the superclass, but for `AssemblyRegionWalker` it was the unit of I/O and `AssemblyRegion` was the unit of processing, and I couldn't update the docs for `ReadWindowWalker` to clear up the confusion without mentioning `AssemblyRegion`-specific concepts.

* The `ReadShard` / `ReadWindow` was/is **only** there to prove that we can shard the data without introducing calling artifacts, and to provide a unit of parallelism for the upcoming Spark implementation. It's not something we really want to expose to users as a prominent knob, and we may hide it completely in the future once the shard size is tuned for performance.

* Inheriting from a more abstract walker type caused a number of other problems as well: methods that should have been final in the supertype could no longer be made final, with the result that tool implementations could inappropriately override engine initialization/shutdown routines. Also, there were issues with the progress meter, since both the supertype traversal and subtype traversal needed their own progress meter for their different units of processing.

Ultimately it was just too awkward and forced, and the read shard is something that we eventually want to make an internal/encapsulated implementation detail anyway. GATK3 made the mistake, I think, of using long, confusing inheritance chains for its walker types, with the result that you got awkward and forced relationships like `RodWalker` inheriting from `LocusWalker`. It's better, I think, to make each traversal as standalone as possible, especially given the simplicity of writing a new walker type in GATK4.

For all of these reasons we don't want `AssemblyRegionWalker` to inherit from a more abstract traversal type -- it's just going to be its own standalone thing, so that it can evolve freely without affecting anyone else. For `SlidingWindowWalker`, which we still want to merge, I recommend making the traversal do **exactly** what you want for your use case, as clearly and simply as possible, without worrying about serving as a base class for other traversals. Ping me once you're happy with it, and I'll re-review.

magicDGS commented:
I agree that it is better to keep it simpler. I was proposing this before looking the latest commits in the HC branch.

I will work on this walker without thinking about other cases, but I would like to keep the idea of padding and slide over intervals instead of the genome from the begining. It will be useful for the things that I have in mind.

Should I close this PR until I implement everything, @droazen?

droazen commented:
@magicDGS You can keep this open -- just push the new version into this same branch and add a comment when you're happy with it and I'll re-review.

magicDGS commented:
I finished the implementation for the draft `SlidingWindowWalker` (I should implement an example and an integration test, but I would like to wait till some issues are solved).

 made a "TODO" about the way in which the intervals are constructed, because I will need a that `ReadShard` have a way to construct a shard without `ReadSource` (either null or empty source), just in case that the implemented `SlidingWindowWalker` does not require reads. @droazen, could you review and give me some feedback about this, because this class is important for other parts of GATK?

magicDGS commented:
@droazen, could you review this PR? Before I implement the integration test I would be important that you check if it is possible the change of the `ReadShard` contract (null `ReadsDataSource` are now allowed, and I don't know if this will affect other parts of GATK.

Thank you very much in advance.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReadShard.java:
```diff
@@ -2,6 +2,7 @@
 
 import htsjdk.samtools.SAMSequenceDictionary;
 import htsjdk.samtools.util.Locatable;
+import org.apache.avro.test.Simple;
```
Remove stray import

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+import java.util.List;
+
+/**
+ * An SlidingWindowWalker is a tool that processes a single window over the genome/intervals at a time, with the ability
```
`An SlidingWindowWalker` -> `A SlidingWindowWalker`

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+ * An SlidingWindowWalker is a tool that processes a single window over the genome/intervals at a time, with the ability
+ * to query optional overlapping sources of reads, reference data, and/or variants/features.
+ *
+ * The windows are constructed over for each interval with a concrete window and step size, and padding overlapping
```
`over for each` -> `over each`

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+
+    protected final static String WINDOW_STEP_ARGUMENT = "windowStep";
+
+    protected final static String WINDOW_PADDING_ARGUMENT = "windowPadding";
```
Can you make these into `@Argument`s here in the base class? GATK traversals generally come with the command-line args needed to control them. Tools could provide their own defaults via an overridable method:

       @Argument(fullName="windowSize", shortName="windowSize", doc = "", optional = true)
       protected int windowSize = defaultWindowSize();

magicDGS replied:
About the arguments, I'm not sure if in this case fit. One of the problems is that some tools implementing a `SlidingWindowWalker` won't like to expose to the final user some parameters. One of the use case that I can imagine is a window walker that does not require any padding; or a walker that requires exactly the same window size and step (as in the case of the `AssemblyRegionWalker`).

Is there a way of hide the `|@Argument` annotations in classes that extends this tool?

droazen replied:
There's currently no way to hide `@Arguments` directly, but you could effectively do this via `ArgumentCollections`. See, eg., the way `GATKTool` uses `ArgumentCollections` to select between required and optional versions of the same arguments (such as `RequiredReadInputArgumentCollection` vs. `OptionalReadInputArgumentCollection`). You could do the same thing in your `SlidingWindowWalker` to allow tool authors to turn different arguments on and off, by having an "empty" argument collection for each sliding window argument.

This is a bit ugly, but I still think it's better to have the arguments controlling traversal defined in the traversal itself rather than in the tools.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+    protected final static String WINDOW_PADDING_ARGUMENT = "windowPadding";
+
+    /**
+     * Get the window size for make the windows
```
`for make` -> `for making`

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+    protected abstract int getWindowSize();
+
+    /**
+     * Get the window step for make the windows
```
Explain what window step means, since it's not obvious

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+        final int windowPadding = getWindowPadding();
+        // either the
+        if ( windowSize <= 1 ) {
+            throw new UserException.BadArgumentValue("--"+WINDOW_SIZE_ARGUMENT, Integer.toString(windowSize), "window size must be > 0");
```
Do you mean `must be > 1` here?

magicDGS replied:
Yes, I do. Thanks for pointing it out.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+            throw new UserException.BadArgumentValue("--"+WINDOW_SIZE_ARGUMENT, Integer.toString(windowSize), "window size must be > 0");
+        }
+        if ( windowStep <= 1 ) {
+            throw new UserException.BadArgumentValue("--"+WINDOW_STEP_ARGUMENT, Integer.toString(windowSize), "window step must be > 0");
```
`must be > 1` ?

magicDGS replied:
Done. Also `windowPadding <= 0` is wrong.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+        final int windowSize = getWindowSize();
+        final int windowStep = getWindowStep();
+        final int windowPadding = getWindowPadding();
+        // either the
```
complete this comment or remove it

magicDGS replied:
Removed

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/SlidingWindowWalker.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    /**
+     * Implementation of window-based traversing. Subclasses can override to provide their own behavior but default
```
`traversing` -> `traversal`

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReadShard.java:
```diff
@@ -36,12 +37,11 @@
      * @param interval the genomic span covered by this shard
      * @param paddedInterval the span covered by this shard, plus any additional padding on each side (must contain the un-padded interval)
-     * @param readsSource source of reads from which to populate this shard
+     * @param readsSource source of reads from which to populate this shard; it could be null if no data source is available
```
I think it's ok if you want to allow `ReadShards` to have a null data source, provided you add a `public boolean hasBackingDataSource()` method that returns false when `readsSource == null` to allow the client to query this.

It seems strange, though, to imagine a `SlidingWindowWalker` that does not take reads, and yet takes `ReadShards` in its `apply()` method. I wonder if perhaps we need a `Shard` superclass for `ReadShard` that contains only `interval` and `paddedInterval`? Then your apply method could take a `Shard, ReadsContext, ReferenceContext, FeatureContext`. Or are you happy with the existing implementation?

magicDGS replied:
I'm happy with the implementation because reads are loaded in a lazy way, and that could be important in terms of performance. Nevertheless, my aim was to provide a base class for two classes of `WindowWalker`: one for reads and one for variants, but allowing to do others on demand for reference, other features, etc.

Thus, it is true that is wierd to iterate over `ReadShard` without reads, only needed to keep track of the `interval`/`paddedInterval` in a single classe. Thus, I'm planning to split the implementation in two to classes (`SlidingWindowReadWalker` and `SlidingWindowVariantWalker`), to keep abstractions simple.

By the way, I implemented for the moment the method that you asked for to keep track of the changes and comments.

droazen commented:
@magicDGS Did a review pass and gave my thoughts on the `ReadShard` issue

magicDGS commented:
Thanks for the help here, @droazen. Now I understand why keeping the walkers separated is a better idea. I will continue working on this soon, separating two `SlidingWindow` for reads and variants, but taking into account your comments.

Do you think that it is better to implement them in two different PRs?

droazen commented:
@magicDGS It might be best to implement both the reads and variants versions in this PR so that we can make sure the interface makes sense before merging, but it's up to you. Let me know when you're ready for another review.

magicDGS commented:
This is ready for review again, @droazen. I split the walkers in two: `ReadSliderWalker` and `VariantSliderWalker`, both based on `ReadWalker` and `VariantWalker` to traverse in a sliding-window approach. I also implemented some `ArgumentCollectionDefinition` for the three parameters (window-size, window-step and window-padding).

There are some changes that I would like to discuss with you:

* `ShardSource<T>`: a generic class for lazily load sources of certain type. Now `ReadShard` extends this class, but there is actually no change in the definitions. I wonder if it is worthy to maintain `ReadShard` as a separate class, because it seems that the only usage is in `AssemblyRegionWalker` and it could be easily changed by a `ShardSource<GATKRead>`.
* `FilteringIterator<T>`: another generic class for filter on iteration records. Now `ReadFilteringIterator` extends this class, but again I don't find any usage that requires an specific case instead of a generic.
* There is some repetition in the test code: `ShardSourceUnitTest` and `ReadShardUnitTest` are exactly the same. I didn't want to remove the later, but I think that it is redundant; the same for the `FilteringIteratorUnitTest` adn `ReadFilteringIterator.
* Because of the inclusion of `ArgumentCollectionDefinition`s for window-related parameters, I just want to know if I should include them in `AssemblyRegionWalker` to maintain consistency in the user options.

Back to you @droazen, and thanks again for all the help.

magicDGS commented:
Can you review when you have time, @droazen? I'll be very thankful if so.

Thanks in advance.

droazen commented:
Yes, apologies for the delay -- we have several quarterly goals that we're trying to get done within the next 2 weeks. I'll try to make time for this review next week.

magicDGS commented:
No rush if you are busy. I've just wanted to highlight that this is waiting for revision. Thanks a lot!

magicDGS commented:
Could it be possible for you to review this if you have time, @droazen?

droazen commented:
I haven't forgotten -- I'll have time to review this after our alpha-2 release is out.

magicDGS commented:
Thanks a lot @droazen!!

davidbernick commented:
Can one of the admins verify this patch?

magicDGS commented:
I've just mess up all the history in this PR, sorry... I will try to solve it.

magicDGS commented:
@davidbernick, I rebased to the current master and solve the conflicts, because the PR was kind of old and new things were implemented (like the downsampler in `ReadShard`.

I think that it is prepared for review, @droazen.

davidbernick commented:
jenkins test please

magicDGS commented:
I extracted some classes to a different branch (#2023) to make it easier to review and integrate with the changes for HaplotypeCallerSpark. I will wait for that branch to be accepted to rebase and update this one...

magicDGS commented:
Hi @droazen, I finished the implementation of the Slider Walkers using the new `Shard` interface from the changes for `HaplotypeCallerSpark` (#2016). I know that it is a lot of code, but I will be very grateful if you can have a look to this PR to accept it.

Thank you in advance!

magicDGS commented:
Hello @droazen, I rebased the PR to the latest master. Could you have a look to this, please? I will be very thankful if so. Thanks in advance!

magicDGS commented:
Sorry @droazen, the previous commit had an error in the tests. I'm rebasing/squashing to make a clear PR and when all check pass (except CLOUD), you can review if you have time. Thank you very much.

magicDGS commented:
Friendly ping here, @droazen! I would appreciate if this could be accepted soon into the framework!

magicDGS commented:
Hello @droazen. Is there any possibility to review this soon? Thanks in advance!

droazen commented:
Sorry @magicDGS, we'll get to this (and your other PRs) as soon as we're able -- once we're out of alpha and the GATK3 team joins us here we should be able to respond to external pull requests much more promptly, but until then we'll unfortunately have to live with slow turnaround time on code reviews.

magicDGS commented:
Thanks for the reply @droazen! I know that it is a lot of work without code that it is not strictly useful within the milestones, and that's why I started this PR time ago (I will need it soon in my work, but I can just use this branch if I rebase from time to time).

I just want to point to this PR again, although I know that it is not trivial code and will take a while to accept it. I'm looking forward for your comments on this. Thanks again!

magicDGS commented:
Any progress/update with respect to this review, @droazen? Thanks in advance!

magicDGS commented on src/test/java/org/broadinstitute/hellbender/cmdline/argumentcollections/WindowPaddingArgumentCollectionTest.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class WindowPaddingArgumentCollectionTest extends BaseTest{
```
Self-reminder: extend `GATKBaseTest` (requires rebase)

magicDGS commented on src/test/java/org/broadinstitute/hellbender/cmdline/argumentcollections/WindowSizeArgumentCollectionTest.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class WindowSizeArgumentCollectionTest extends BaseTest{
```
Self-reminder: extend `GATKBaseTest` (requires rebase)

magicDGS commented on src/test/java/org/broadinstitute/hellbender/cmdline/argumentcollections/WindowStepArgumentCollectionTest.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class WindowStepArgumentCollectionTest extends BaseTest{
```
Self-reminder: extend `GATKBaseTest` (requires rebase)

magicDGS commented on src/test/java/org/broadinstitute/hellbender/utils/iterators/FilteringIteratorUnitTest.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Iterator;
+import java.util.function.Predicate;
+
+public class FilteringIteratorUnitTest extends BaseTest {
```
Self-reminder: extend `GATKBaseTest` (requires rebase)

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/LocalReadShard.java:
```diff
@@ -12,10 +11,7 @@
-import java.util.stream.StreamSupport;
 
 /**
  * A class to represent a shard of reads data, optionally expanded by a configurable amount of padded data.
```
@droazen - looking into the resurrection of this branch, I realized that the `LocalReadShard` disappeared. For the sliding-window walker I required that each window is independent of the others, so I guess that the `MultiIntervalLocalReadShard` is not the option to go.

Is it ok to keep the `LocalReadShard`? Otherwise, I think that starting a new project for not keeping the GATK blow with unnecessary walkers for your methods is the best way to go...

droazen replied:
@magicDGS I think it would be ok to resurrect `LocalReadShard` if you really need it. But you should know that the reason we got rid of it was because we found that there are major performance issues with querying each interval individually. It's much more efficient to query a large number of intervals at once, particularly if the intervals are close together or overlap, to avoid reading and parsing the same parts of the bam file multiple times. Once we moved to `MultiIntervalLocalReadShard`, the performance of both the `HaplotypeCaller` and `Mutect2` increased by about 30-40%.

I'd recommend closing out these old PRs and submitting a new PR for `SlidingWindowWalker` with the walker class itself, an `ExampleSlidingWindowWalker` and `ExampleSlidingWindowWalker`, and whatever utilities you require (but note that the less common code you refactor, the easier it will be for us to merge your PR quickly).

magicDGS commented:
Closing in favor of a new implementation
